src/guide-rngs.md

```diff
@@ -6,8 +6,314 @@ however, often you can just use [`thread_rng`]. This function
+with a partially overlapping slice of the stream.
+
+If the period of the RNG is 2<sup>128</sup>, and an application consumes
+2<sup>48</sup> values, it then takes about 2<sup>32</sup> random
+initializations to have a chance of 1 in a million to repeat part of an
+already used stream. This seems good enough for common usage of
+non-cryptographic generators, hence the recommendation of at least
+2<sup>128</sup>. As an estimate, the chance of any overlap in a period of
+size `p` with `n` independent seeds and `u` values used per seed is
+approximately `1 - e^(-u * n^2 / (2 * p))`.
+
+Further, it is not recommended to use the full period of an RNG. Many
+PRNGs have a property called *k-dimensional equidistribution*, meaning that
+for values of some size (potentially larger than the output size), all
+possible values are produced the same number of times over the generator's
+period. This is not a property of true randomness. This is known as the
+generalized birthday problem, see the [PCG paper] for a good explanation.
+This results in a noticable bias on output after generating more values
+than the square root of the period (after 2<sup>64</sup> values for a
+period of 2<sup>128</sup>).
+
+
+# Security
+
+## Predictability
+
+From the context of any PRNG, one can ask the question *given some previous
+output from the PRNG, is it possible to predict the next output value?*
+This is an important property in any situation where there might be an
+adversary.
+
+Regular PRNGs tend to be predictable, although with varying difficulty. In
+some cases prediction is trivial, for example plain Xorshift outputs part of
+its state without mutation, and prediction is as simple as seeding a new
+Xorshift generator from four `u32` outputs. Other generators, like
+[PCG](http://www.pcg-random.org/predictability.html) and truncated Xorshift*
+are harder to predict, but not outside the realm of common mathematics and a
+desktop PC.
+
+The basic security that CSPRNGs must provide is the infeasibility to predict
+output. This requirement is formalized as the [next-bit test]; this is
+roughly stated as: given the first *k* bits of a random sequence, the
+sequence satisfies the next-bit test if there is no algorithm able to
+predict the next bit using reasonable computing power.
+
+A further security that *some* CSPRNGs provide is forward secrecy:
+in the event that the CSPRNGs state is revealed at some point, it must be
+infeasible to reconstruct previous states or output. Note that many CSPRNGs
+*do not* have forward secrecy in their usual formulations.
+
+As an outsider it is hard to get a good idea about the security of an
+algorithm. People in the field of cryptography spend a lot of effort
+analyzing existing designs, and what was once considered good may now turn
+out to be weaker. Generally it is best to use algorithms well-analyzed by
+experts, such as those recommended by NIST or ECRYPT.
```

Comment:
Maybe add links? I hesitate to defer to NIST after the Dual_EC_DRBG disaster.

Reply:
I hesitate to fully trust any single authority, but it's really beyond my expertise to give better advice here. Any suggestions?